Rewrite CreateNewBlock

morcos opened the pull request:
WARNING: This hasn't been used for mainnet mining yet as far as I know.

~~NOTE: This includes a commit which changes the default block priority size to 0.  This code has been optimized for not including transactions in a priority space.  If block priority space > 0 then the code does not offer too much of a performance improvement over the existing code.~~ (EDIT: now its fast regardless)

The mempool is explicitly assumed to be responsible for maintaining consistency of transactions with respect to not spending non-existent outputs, not double spending, script validity and coinbase maturity.  Only finality of transactions is checked before assembling a block.

A final call to TestBlockValidity is still performed.  If an invalid block has been constructed, then an error is thrown ~~logged and a NULL pointer is returned.~~ (EDIT: reverted to original behavior with more informative error)

This code produces the same blocks as the code it replaces with the following exceptions:
- The fee rate sort tie breaker is hash instead of priority
- The priority sort has a secondary tie breaker of hash
- ~~The priority block is not limited to transactons above AllowFree()~~ (EDIT: fixed)
- ~~The fee rate sorted part of the block is not limited to transactions above minRelayTxFee~~ (EDIT: fixed)
- (EDIT: The blocks are the same if you keep scanning to the end to try to fill up the last remaining space, but this code stops after trying 50 times to fill the last 1000 bytes)

Comparing this to the original code over 2000 calls to CreateNewBlock over the last 2 days.
(blockprioritysize=0, maxmempool=300, dbcache=1000, maxsigcachesize=1000000)
Time in ms

|  | Time to assemble block | Additional time to perform final TestBlockValidity |
| --- | --- | --- |
| master | 2602 | 240 |
| this pull | 14 | 438 |

The extra slowness in TestBlockValidity is because the cache used to be warmed up in the assembly code.

morcos committed:
Store the total sig op count of a tx.

Store sum of legacy and P2SH sig op counts.  This is calculated in AcceptToMemory pool and storing it saves redoing the expensive calculation in block template creation.

morcos committed:
Add a score index to the mempool.

The score index is meant to represent the order of priority for being included in a block for miners.  Initially this is set to the transactions modified (by any feeDelta) fee rate.  Index improvements and unit tests by sdaftuar.

morcos committed:
Make accessing mempool parents and children public

morcos committed:
Expose FormatStateMessage

luke-jr commented on src/txmempool.h:
```diff
@@ -78,7 +80,8 @@ class CTxMemPoolEntry
     CTxMemPoolEntry(const CTransaction& _tx, const CAmount& _nFee,
-                    int64_t _nTime, double _dPriority, unsigned int _nHeight, bool poolHasNoInputsOf = false);
+                    int64_t _nTime, double _dPriority, unsigned int _nHeight, bool poolHasNoInputsOf = false,
+                    unsigned int nSigOps = 0);
```
NACK defaulting nSigOps to 0 when we depend on it being accurate.

morcos replied:
ok i agree, it makes more sense to default it high.

luke-jr replied:
It does not make sense to default it period. Note the exact sigop count is part of the GBT response.

morcos replied:
OK agreed.  The default is just for the tests, I think for now it makes sense to remove all defaults from the constructor and just make the tests fill them in.  Unfortunately there are 2 other pulls open which also add arguments to the constructor and require changing the tests.  I'm going to hold off updating the tests in this pull until we're a bit closer assuming those might be merged first.

luke-jr commented on src/miner.cpp:
```diff
@@ -351,8 +296,10 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
-            throw std::runtime_error("CreateNewBlock(): TestBlockValidity failed");
+        if (!TestBlockValidity(state, *pblock, pindexPrev, false, false)) {
+            LogPrintf("ERROR: CNB TestBlockValidity failed: %s\n",FormatStateMessage(state));
+            return NULL;
```
Why change exception to returning NULL here? IMO this should be an independent PR.

morcos replied:
The idea was this is now more likely to happen, and it seemed to meet that the JSON error from GetBlockTemplate would still be sufficient notification, and it would be better to not bring the node down.  However, I'm happy to leave the existing behavior if people think crashing the node is the right outcome.

luke-jr replied:
Exceptions here should not crash the node... it just gets returned as a nicer JSON-RPC error.

morcos committed:
Add TxPriority class and comparator

morcos committed:
Rewrite CreateNewBlock

Use the score index on the mempool to only add sorted txs in order.  Remove much of the validation while building the block, relying on mempool to be consistent and only contain txs that can be mined.
The mempool is assumed to be consistent as far as not containing txs which spend non-existent outputs or double spends, and scripts are valid.  Finality of txs is still checked (except not coinbase maturity, assumed in mempool).
Still TestBlockValidity in case mempool consistency breaks and return error state if an invalid block was created.
Unit tests are modified to realize that invalid blocks can now be constructed if the mempool breaks its consistency assumptions and also updated to have the right fees, since the cached value is now used for block construction.

Conflicts:
	src/miner.cpp

morcos commented:
OK, this has been rewritten to use `CTxMemPoolEntry::GetPriority` for the priority sort.
This will make it very fast even when maintaining a priority space in the block.  So the commit changing the default blockprioritysize has been removed.  

Of course until #6357 is merged or some other decision is taken regarding priority, it will not sort by an accurate measure of priority.

morcos commented:
I addressed @luke-jr 's comment about changing the method of failure return.  I still modified it to give some extra state info.   That commit and other fixes will get squashed.

Among the remaining outstanding items is to fix the constructor for CTxMemPoolEntry to not have any default arguments.  It would be nice to know a merge order since this will conflict with #6357 and #6915 and involves tedious updating of a bunch of tests.

morcos commented:
I rebased this and squashed fixes for comments so far.

I rebased it on top of #7008, because I needed to pick some path forward on how priority would be addressed in this PR.  The first 3 commits are therefore part of that pull and not this one.  If we decide to merge #6357 as well, then it should be an easy change.

I will continue testing, but I don't have any other changes planned right now.

morcos commented:
Thanks @sdaftuar for finding a bug in the way I was doing PrioritiseTransaction.   There should probably be RPC tests for those functions.

morcos commented:
Apologies for all the churn.  Squashed again and rebased off a slightly modified index @sdaftuar created for #7062.   This introduces one more minor difference from the original code in that fee rates are compared as doubles now.  

If I correct for these differences:
- using a score index which compares fee rates not doubles
- make the original code use the hash tie breakers
- search up 5000 txs to fill the final 1000 bytes
- implement #6357 to correctly compute priority

Then I generate the exact same blocks as the old code modulo a few rare cases where double imprecision in the different priority calculation causes tx reordering in the priority space.

jtimon commented on src/miner.cpp:
```diff
@@ -350,8 +287,9 @@ CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const CScript& s
-        if (!TestBlockValidity(state, chainparams, *pblock, pindexPrev, false, false))
-            throw std::runtime_error("CreateNewBlock(): TestBlockValidity failed");
+        if (!TestBlockValidity(state, chainparams, *pblock, pindexPrev, false, false)) {
+            throw std::runtime_error("CreateNewBlock(): TestBlockValidity failed: " + FormatStateMessage(state));
```
can you ```strprintf("%: TestBlockValidity failed: %", __func__, ...``` instead?

sdaftuar commented:
Addressed @jtimon's nit.

btcdrak commented:
@morcos Needs rebase.

jameshilliard commented:
TestBlockValidity should probably be done in a background thread so that it does not delay GBT since TestBlockValidity is just a sanity check.

morcos commented:
rebased

gmaxwell commented:
Concept ACK. Some tested-ack. Will test more now.

jgarzik commented:
ut ACK

laanwj commented:
Needs rebase again...

morcos commented:
that was an expected merge conflict, will rebase shortly

@laanwj my warning about not having mined, was just a literal warning because i don't have sufficient hashpower to mine on mainnet.  I have created 1000s upon 1000s of block templates that all pass TestBlockValidity both in code running this pull live for the past 2 weeks (and prior iterations for weeks before that) and in historical simulation over 6 weeks of data.

sipa commented:
ACK

morcos commented:
rebased

luke-jr commented:
This was not ready for merging. It needs #6357 first.

laanwj commented:
> This was not ready for merging. It needs #6357 first.

#7008, which took the place of #6537 was merged
